Fixed #31594 -- Added ASGIStaticFilesHandler.get_response_async().

massover opened the pull request:
This code comes from conversions on this [thread](https://forum.djangoproject.com/t/current-async-todos/1013/13) on the django forum. While trying to add asgi support to runserver, I noticed that the get_response_async method was missing.

felixxm commented:
@massover Thanks, please create [a new ticket](https://code.djangoproject.com/newticket).

massover commented:
@felixxm https://code.djangoproject.com/ticket/31594

Let me know if there's anything else to do.

charettes commented:
@massover I think the issue is bit more tricky here `self.serve` ultimately calls `django.contrib.staticfiles.views.serve` and `django.views.static.service` which both do synchronous I/O

https://github.com/django/django/blob/94f63b926fd32d7a7b6e2591ef72aa8f040f25cc/django/contrib/staticfiles/views.py#L33

https://github.com/django/django/blob/94f63b926fd32d7a7b6e2591ef72aa8f040f25cc/django/views/static.py#L41

https://github.com/django/django/blob/94f63b926fd32d7a7b6e2591ef72aa8f040f25cc/django/views/static.py#L44

https://github.com/django/django/blob/94f63b926fd32d7a7b6e2591ef72aa8f040f25cc/django/views/static.py#L50

Implementing `get_response_async` will involve making all the call chain `async` as well.

massover commented:
@charettes ah good point. Without any native async file io, seems like we should wrap it with sync_to_async?

charettes commented:
@massover I guess so, IIRC it relies on a thread pool.

massover commented on tests/handlers/tests.py:
```diff
@@ -88,6 +91,17 @@ def test_invalid_multipart_boundary(self):
+
+    @modify_settings(INSTALLED_APPS={'append': 'staticfiles_tests.apps.test'})
+    async def test_get_async_response(self):
+        request = self.request_factory.post('/static/test/file.txt')
```
Oops, yeah, was a copy paste error on a post. Didn't know about the AsyncRequestFactory. Thanks!

massover commented on tests/handlers/tests.py:
```diff
@@ -88,6 +91,17 @@ def test_invalid_multipart_boundary(self):
+class TestASGIStaticFilesHandler(SimpleTestCase):
+    request_factory = RequestFactory()
+
+    @modify_settings(INSTALLED_APPS={'append': 'staticfiles_tests.apps.test'})
```
Can you confirm what "that" is? If you mean doing `modify_settings`, without the decorator I end up 404'ing, presumably because it's not included in `INSTALLED_APPS`

massover commented on tests/handlers/tests.py:
```diff
@@ -88,6 +91,17 @@ def test_invalid_multipart_boundary(self):
         self.assertEqual(response.status_code, 400)
 
 
+class TestASGIStaticFilesHandler(SimpleTestCase):
```
👍 moved

massover commented on tests/staticfiles_tests/test_handlers.py:
```diff
@@ -0,0 +1,15 @@
+        request = self.async_request_factory.get('/static/test/file.txt')
+        handler = ASGIStaticFilesHandler(ASGIHandler())
+        response = await handler.get_response_async(request)
+        response.close()
```
Just a note: without `response.close()` here, the test ends up with the following:

```
test_get_async_response (tests.staticfiles_tests.test_handlers.TestASGIStaticFilesHandler) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.005s

OK
Exception ignored in: <_io.FileIO name='/Users/joshmassover/Development/django/tests/staticfiles_tests/apps/test/static/test/file.txt' mode='rb' closefd=True>
ResourceWarning: unclosed file <_io.FileIO name='/Users/joshmassover/Development/django/tests/staticfiles_tests/apps/test/static/test/file.txt' mode='rb' closefd=True>
```

felixxm commented on django/contrib/staticfiles/handlers.py:
```diff
@@ -52,6 +54,12 @@ def get_response(self, request):
+        try:
+            return await sync_to_async(self.serve)(request)
+        except Http404 as e:
+            return response_for_exception(request, e)
```
`return await sync_to_async(response_for_exception)(request, e)`?

felixxm replied:
Can we add tests for catching 404s?

carltongibson replied:
We wrap `response_for_exception` in an async context elsewhere:

https://github.com/django/django/blob/f47d5aac622c334ebeba06b7460204aeb98661e2/django/core/handlers/exception.py#L34-L51

carltongibson reviewed (commented):
Hi @massover. Thanks for this! 

I squashed, and rebased, and adjusted the [commit message format](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines).

The addition of the `test_handlers.py` file is fine: none of the other files look a natural home. 
I adjusted the test case to subclass `StaticFilesTestCase`, which addresses Simon's point about needing to modify the settings. 

Beyond @felixxm's comments, I think we could do with a (single) test using `ApplicationCommunicator` in `django/tests/asgi/tests.py`, showing that the full application does correctly serve. (That's the test that would fail on master right?) I'd be happy to add it but I thought you might want to do that? 

~~This strikes me as a bug in a new feature in 3.0. (Discuss 🙂)~~ Forget that last bit: the `_async` path doesn't exist in 3.0.

massover commented:
> Hi @massover. Thanks for this!
> 
> I squashed, and rebased, and adjusted the [commit message format](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines).
> 
> The addition of the `test_handlers.py` file is fine: none of the other files look a natural home.
> I adjusted the test case to subclass `StaticFilesTestCase`, which addresses Simon's point about needing to modify the settings.
> 
> Beyond @felixxm's comments, I think we could do with a (single) test using `ApplicationCommunicator` in `django/tests/asgi/tests.py`, showing that the full application does correctly serve. (That's the test that would fail on master right?) I'd be happy to add it but I thought you might want to do that?
> 
> ~This strikes me as a bug in a new feature in 3.0. (Discuss 🙂)~ Forgot that last bit: the `_async` path doesn't exist in 3.0.

Thanks @carltongibson! I'll finish up those tests as per your request and @felixxm

felixxm commented:
@massover Thanks :+1: Welcome aboard :boat: 

I simplified tests.
